csharp/src/SeedLang.Block/Position.cs

```diff
@@ -0,0 +1,60 @@
+// Copyright 2021 The Aha001 Team.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using SeedLang.Common;
+
+namespace SeedLang.Block {
+  // The position where a block docks to another block, or stays as a standalone block on a
+  // SeedBlock module canvas.
+  //
+  // This class is different with SeedLang.Common.BlockPosition, which represents a syntaxical code
+  // position in a SeedBlock module.
+  public class Position {
+    // How the block is docked to another block.
+    public enum DockType {
+      // The block is standalone, not docked to any other block. The block's position is defined by
+      // CanvasPosition.
+      UnDocked = 0,
+      // The block docks as one of the target block's input blocks. The index of the input position
+      // is specified by DockPosition.
+      Input,
```

Comment:
The term "input" comes from Google Blockly's terminology, with a slight modification.

See https://developers.google.com/blockly/guides/create-custom-blocks/define-blocks#block_inputs for Google Blockly's definition.

In Google Blockly, "input" means all types of blocks that can be connected to the target block.

Compared to Google Blockly, SeedLang Block has some significant design decisions:

* There is no "output" connection in blocks. I.e., all blocks are divided into two categories: statements (without result value) and expressions (with result values). With this design, there is no need to specify how it can be connected to another block's input slot. Thus in SeedLang Block, every input slot can be docked with an expression block, taking its result value as the input value.
* Based on the above consideration, it's required to distinguish the Input slots from the statement slots.

I will describe this in detail in the design doc of SeedBlock, at /design/seed_block.md shortly. Will open an issue for me on this task.

Reply:
So Google Blocky's definition is all the connected blocks are inputs.
We will modify it to:
- An input slot means the slot that can be docked with an expression block. So the input means this block can provide an input to the parent block. Is it right?
- Other slots are Statement slots which is used to connect statements.

For other developers, they just need to think about and understand the meaning of the "input".

Reply:
That's right. But the term "input" here has broader meanings than in a language's grammar. E.g., considering a *function def* block, it will have 1+N input slots and 1 statement slot:

* 1 Input slot to define the function name: an identifier block can be docked there.
* N Input slots to define the parameters of the function: N identifier blocks can be docked there.
* 1 Statement slot to define the body of the function.

I will provide a clear description, as well as figures in the design doc.